File: daily_python/85_Maximal_Rectangle.py
import os
import sys
from typing import List

class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        m, n = len(matrix), len(matrix[0])
        ans = 0
        dp = [0] * n
        for i in range(m):
            temp = [0] * n
            st = []
            for j in range(n + 1):
                if j == n:
                    currH = 0
                else:
                    currH = dp[j] + 1 if matrix[i][j] == '1' else 0
                    temp[j] = currH
                while st and temp[st[-1]] >= currH:
                    h = temp[st.pop()]
                    leftSmaller = st[-1] if st else -1
                    ans = max(ans, h * (j - leftSmaller - 1))
                st.append(j)
            dp = temp
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            matrix = [list(map(str, row.split("','"))) for row in lines[0].split('=')[1].strip()[3:-4].split("'},{'")]
            dataset.append(matrix)
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for matrix in dataset:
        results.append(solution.maximalRectangle(matrix))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2026/85_Maximal_Rectangle.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

// class Solution {
// public:
//     int maximalRectangle(vector<vector<char>>& matrix) {
//         int m = matrix.size();
//         int n = matrix[0].size();
//         vector<vector<int>> dp(m, vector<int>(n, 0));
//         for (int i = 0; i < m; i++){
//             if (matrix[i][0] == '1') dp[i][0] = 1;
//             for (int j = 1; j < n; j++){
//                 if (matrix[i][j] == '1') dp[i][j] = 1 + dp[i][j-1];
//                 else dp[i][j] = 0;
//             }
//         }
//         int ans = 0;
//         for (int i = 0; i < m; i++){
//             for (int j = 0; j < n; j++){
//                 if (matrix[i][j] == '0') continue;
//                 int curr = dp[i][j];
//                 for (int k = i; k >= 0; k--){
//                     curr = min(curr, dp[k][j]);
//                     ans = max(ans, curr * (i - k + 1));
//                 }
//             }
//         }
//         return ans;
//     }
// };

class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        int ans = 0;
        vector<int> dp(n, 0);
        for (int i = 0; i < m; i++){
            vector<int> temp(n, 0);
            stack<int> sta;
            for (int j = 0; j <= n; j++){
                int curr;
                if (j == n) curr = 0;
                else{
                    curr = (matrix[i][j] == '0')? 0 : dp[j] + 1;
                    temp[j] = curr;
                }
                while(!sta.empty() && temp[sta.top()] >= curr){
                    int h = temp[sta.top()];
                    sta.pop();
                    int leftSmaller = (sta.empty())? -1 : sta.top();
                    ans = max(ans, h * (j - leftSmaller - 1));
                }
                sta.push(j);
            }
            swap(dp, temp);
        }
        return ans;
    }
};

int main(){
    vector<vector<char>> matrix = {{'1','0','1','0','0'},{'1','0','1','1','1'},{'1','1','1','1','1'},{'1','0','0','1','0'}};
    // vector<vector<char>> matrix = {{'0'}};
    // vector<vector<char>> matrix = {{'1'}};
    Solution S;
    cout << S.maximalRectangle(matrix) << endl;
    return 0;
}